kedro/pipeline/modular_pipeline.py

```diff
@@ -64,16 +65,24 @@ def _validate_datasets_exist(
     parameters: AbstractSet[str],
     pipe: Pipeline,
 ) -> None:
+    """Validate that inputs, parameters and outputs map correctly onto the provided nodes."""
     inputs = {_strip_transcoding(k) for k in inputs}
     outputs = {_strip_transcoding(k) for k in outputs}
 
     existing = {_strip_transcoding(ds) for ds in pipe.datasets()}
     non_existent = (inputs | outputs | parameters) - existing
     if non_existent:
-        raise ModularPipelineError(
-            f"Failed to map datasets and/or parameters: "
-            f"{', '.join(sorted(non_existent))}"
+        possible_matches = []
+        for non_existent_input in non_existent:
+            possible_matches += difflib.get_close_matches(non_existent_input, existing)
+
+        error_msg = f"Failed to map datasets and/or parameters onto the nodes provided: {', '.join(sorted(non_existent))}"
+        suggestions = (
+            f" - did you mean one of these instead: {', '.join(sorted(possible_matches))}"
```

Comment:
Sorting sometimes breaks the order so that the order `non_existent` differs from `possible_matches` found when there is more than one `non_existent` input. For example:
```
def create_pipeline(**kwargs) -> Pipeline:
    pipeline_instance = pipeline(
        [
            node(
                func=split_data,
                # Note wrong input name
                inputs=["model_input_table_NOT_FOUND", "params:model_options"],
                outputs=["X_train_test", "X_test", "y_train", "y_test"],
                name="split_data_node",
            ),
            node(
                func=train_model,
                inputs=["X_train", "y_train"],
                outputs="regressor",
                name="train_model_node",
            ),
            node(
                func=evaluate_model,
                inputs=["regressor", "X_test", "y_test"],
                outputs=None,
                name="evaluate_model_node",
            ),
        ]
    )

    ds_pipeline_1 = pipeline(
        pipe=pipeline_instance,
        inputs=["model_input_table", "X_train_mod"],
        namespace="active_modelling_pipeline",
    )
    ds_pipeline_2 = pipeline(
        pipe=pipeline_instance,
        inputs=["model_input_table", "X_train_mod"],
        namespace="candidate_modelling_pipeline",
    )

    return ds_pipeline_1 + ds_pipeline_2
```

will result this message: `kedro.pipeline.modular_pipeline.ModularPipelineError: Failed to map datasets and/or parameters onto the nodes provided: X_train_mod, model_input_table - did you mean one of these instead: X_train, X_train_test, model_input_table_NOT_FOUND, y_train`

It's not a big problem, but maybe it makes sense to keep grouping order from `get_close_matches`? For example:

```
        for non_existent_input in non_existent:
            possible_matches.append(f"[{', '.join(difflib.get_close_matches(non_existent_input, existing))}]")

        error_msg = f"Failed to map datasets and/or parameters onto the nodes provided: {', '.join(non_existent)}"
        suggestions = (
            f" - did you mean one of these instead: {', '.join(possible_matches)}"
            if possible_matches
            else ""
        )
```

will give a slightly better message: `kedro.pipeline.modular_pipeline.ModularPipelineError: Failed to map datasets and/or parameters onto the nodes provided: X_train_mod, model_input_table - did you mean one of these instead: [X_train, X_train_test, y_train], [model_input_table_NOT_FOUND]`

Wdyt?  🙂

Reply:
Great idea!